### PR TITLE
#122 - Adding lstrip, rstrip, and strip for iterables

### PR DESCRIPTION
**Pull request by mblahay**
Adding lstrip, rstrip, and strip for iterables

### CODE REVIEW

**Comment by mahmoud** [comment]
Cool, I'm generally +1 on this, but I'll defer to @kurtbrose somewhat since it was his idea. If he doesn't chime in, I'll come back through and give it a closer look. Thanks!

**Comment by mblahay** [comment]
@mahmoud, I haven't heard from @kurtbrose nor do I see any recent activity on his account. How long do you want to wait?

**Comment by mblahay** [comment]
@mahmoud, what is the status of this?

**Comment by mblahay** [comment]
poke

**Comment by mahmoud** [comment]
Yeah I don't know where he's at, busy presumably. I'll do a preliminary review.

File: boltons/iterutils.py [hunk 1]
```diff
@@ -177,6 +178,100 @@ def split_iter(src, sep=None, maxsplit=None):
...
+def lstrip(iterable, strip_value=None):
+    """Strips values from the beginning of an iterable. Stripped items will
+    match the value of the argument strip_value. Functionality is analigous
+    to that of the method str.lstrip. Returns a list.
+
+    >>> lstrip(['Foo', 'Bar', 'Bam'], 'Foo')
+    ['Bar', 'Bam']
+
+    """
+    return list(lstrip_iter(iterable, strip_value))
```
Comment (mahmoud) [inline boltons/iterutils.py:190]: One reason I was hoping to get @kurtbrose in here was to ask whether he would expect lstrip to dynamically handle the string case. I.e., autojoin things back together as "abc".lstrip('a') returns a string, not a list of `['b', 'c']`.

File: boltons/iterutils.py [hunk 2]
```diff
@@ -177,6 +178,100 @@ def split_iter(src, sep=None, maxsplit=None):
...
+def strip_iter(iterable,strip_value=None):
+    """Strips values from the beginning and end of an iterable. Stripped items
+    will match the value of the argument strip_value. Functionality is 
+    analigous to that of the method str.strip. Returns a generator.
+
+    >>> list(strip_iter(['Fu', 'Foo', 'Bar', 'Bam', 'Fu'], 'Fu'))
+    ['Foo', 'Bar', 'Bam']
+
+    """
+    return rstrip_iter(lstrip_iter(iterable,strip_value),strip_value)
```
Comment (mahmoud) [inline boltons/iterutils.py:272]: lstrip_iter and rstrip_iter I get, but just plain strip_iter? besides for consistency sake, how will this perform any differently than plain strip()?

File: boltons/iterutils.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 import socket
 import hashlib
 import itertools
+import sys
```
Comment (mahmoud) [inline boltons/iterutils.py:21]: is this import used?

File: tests/test_iterutils.py [hunk 4]
```diff
@@ -509,3 +509,21 @@ def test_chunked_bytes():
     from boltons.iterutils import chunked
 
     assert chunked(b'123', 2) in (['12', '3'], [b'12', b'3'])
+
+
+def test_lstrip():
+    from boltons.iterutils import lstrip
+
+    assert lstrip([0,1,0,2,0,3,0],0) == [1,0,2,0,3,0]
```
Comment (mahmoud) [inline tests/test_iterutils.py:517]: should we test some more edge cases on these?

File: boltons/iterutils.py (hunk 1 above)
Comment (mblahay) [reply boltons/iterutils.py:190]: I wondered about that too but decided against that because it would be a special case to be used for strings only. These changes were meant to generically handle iterables, a class of value that strings happen to be part of. Besides, if this version of the function is problematic for a developer who is working with strings, there are always the string based methods.

File: boltons/iterutils.py (hunk 2 above)
Comment (mblahay) [reply boltons/iterutils.py:272]: I added strip because the standard String class includes lstrip, rstrip, and strip. I reasoned that since the "strip" functionality was to be emulated then all three needed to be available. Of course in reality strip as I coded it is it just joining lstrip and rstrip together to strip from both ends of the iterator.

As for your question about how strip_iter performs against "plain strip()." Not sure I understand. As was the pattern, all of the _iter functions are generators, while the non _iter functions return actual lists. From the standpoint of performance, if you are working with data of unknown size or very large amounts of data, you probably want to use the generators/iterators.

**Comment by mblahay** [comment]
@mahmoud , is this ready for merging?

**Comment by mahmoud** [comment]
@mblahay I think the short answer is: yes :) I'm cutting a release tonight and I figured I'd make the call. Thanks again for all your work and patience!
